### PR TITLE
[Refactor] Rotated FCOS

### PR DESCRIPTION
**Pull request by liuyanyi**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Compared with 0.x, Rotated FCOSHead and CSLFCOSHead merged together.
Added a argument `use_hbbox_loss` to replace `separate_angle` to make it better understanding.

Before:
Arg `separate_angle` control the type of bbox_loss used and whether angle loss is enabled.

After:
Arg `use_hbbox_loss` control the type of bbox_loss. Whether angle loss is enabled is controlled by `loss_angle`. 
In this case we can use Rotated IoU Loss and L1Loss together, Rotated IoU Loss for bbox, L1 for angle auxiliary losses.

Results:
| model | Reported ap50 | map | ap50 | ap75 |
| :---: | :---: | :---: | :---: | :---: |
| Rotated FCOS | 71.28 | 41.10 | 70.87 | 41.21 |
| Rotated FCOS kld | 71.89 | 40.37 | 71.55 | 38.92 |
| Rotated FCOS hbb csl | 71.76 | 40.22 | 70.79 | 39.26 |
| Rotated FCOS hbb | 70.70 | 38.06 | 69.71 | 34.81 |
| Rotated FCOS + l1 | - | 40.90 | 71.18 | 40.87 |

Note: The result is not stable and may have 1% ap difference. Training process may have unstable too.

We should do some exps on tricks and losses to find the best config settings.

Add l1 loss for angle, results does not appear to have improved, but it has not decreased either.

## Modification

- RotatedFCOSHead
- PseudoAngleCoder
- Fix `F.cosine_similarity` typo in oriented_reppoints
- Fix typo of `encode_size`
- Fix CSLCoder empty input when keepdim=True

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmrotate/models/dense_heads/rotated_fcos_head.py [hunk 1]
```diff
@@ -313,24 +257,24 @@ def loss(self,
...
-            tuple:
-                concat_lvl_labels (list[Tensor]): Labels of each level. \
-                concat_lvl_bbox_targets (list[Tensor]): BBox targets of each \
-                    level.
-                concat_lvl_angle_targets (list[Tensor]): Angle targets of \
-                    each level.
+            tuple: Targets of each level.
+            - concat_lvl_labels (list[Tensor]): Labels of each level.
+            - concat_lvl_bbox_targets (list[Tensor]): BBox targets of each \
+            level.
```
Comment (zytx121) [inline mmrotate/models/dense_heads/rotated_fcos_head.py:275]:
```suggestion
              level.
```

File: mmrotate/models/dense_heads/rotated_fcos_head.py [hunk 2]
```diff
@@ -313,24 +257,24 @@ def loss(self,
...
+            - concat_lvl_angle_targets (list[Tensor]): Angle targets of \
+            each level.
```
Comment (zytx121) [inline mmrotate/models/dense_heads/rotated_fcos_head.py:277]:
```suggestion
              each level.
```

File: mmrotate/models/dense_heads/rotated_fcos_head.py [hunk 3]
```diff
@@ -60,92 +46,37 @@ class RotatedFCOSHead(RotatedAnchorFreeHead):
...
-            loss_cls=loss_cls,
-            loss_bbox=loss_bbox,
-            norm_cfg=norm_cfg,
-            init_cfg=init_cfg,
-            **kwargs)
-        self.loss_centerness = TASK_UTILS.build(loss_centerness)
-        if self.separate_angle:
-            self.loss_angle = TASK_UTILS.build(loss_angle)
+        self.angle_coder = TASK_UTILS.build(angle_coder)
+        super(RotatedFCOSHead, self).__init__(**kwargs)
```
Comment (zytx121) [inline mmrotate/models/dense_heads/rotated_fcos_head.py:61]:
```suggestion
        super().__init__(**kwargs)
```

File: mmrotate/models/dense_heads/rotated_fcos_head.py [hunk 4]
```diff
@@ -1,57 +1,43 @@
...
-        norm_on_bbox (bool): If true, normalize the regression targets
-            with FPN strides. Default: False.
-        centerness_on_reg (bool): If true, position centerness on the
-            regress branch. Please refer to https://github.com/tianzhi0549/FCOS/issues/89#issuecomment-516877042.
-            Default: False.
-        separate_angle (bool): If true, angle prediction is separated from
-            bbox regression loss. Default: False.
+        angle_version (str): Angle representations. Defaults to 'le90'.
+        use_hbbox_loss (bool): If true, use horizontal bbox loss and
+            loss_angle should not be None. Default: False.
```
Comment (zytx121) [inline mmrotate/models/dense_heads/rotated_fcos_head.py:34]:
```suggestion
            loss_angle should not be None. Defaults to False.
```

File: mmrotate/models/dense_heads/rotated_fcos_head.py [hunk 5]
```diff
@@ -1,57 +1,43 @@
...
         scale_angle (bool): If true, add scale to angle pred branch. Default: True.
```
Comment (zytx121) [inline mmrotate/models/dense_heads/rotated_fcos_head.py:35]:
```suggestion
        scale_angle (bool): If true, add scale to angle pred branch. Defaults to True.
```

**Review by zytx121 — DISMISSED** [review]
LGTM

File: mmrotate/models/detectors/rotated_fcos.py [hunk 6]
```diff
@@ -1,22 +1,30 @@
...
+from mmdet.models.detectors.single_stage import SingleStageDetector
+from mmdet.utils import ConfigType, OptConfigType, OptMultiConfig
 
+from mmrotate.registry import MODELS
 
-@ROTATED_DETECTORS.register_module()
-class RotatedFCOS(RotatedSingleStageDetector):
+
+@MODELS.register_module()
+class RotatedFCOS(SingleStageDetector):
```
Comment (RangiLyu) [inline mmrotate/models/detectors/rotated_fcos.py:9]: Do we still need this class?

Comment (RangiLyu) [reply mmrotate/models/detectors/rotated_fcos.py:9]: THe configs are using "mmdet.FCOS"

Comment (liuyanyi) [reply mmrotate/models/detectors/rotated_fcos.py:9]: This class is useless, should we remove that?

**Review by RangiLyu — APPROVED** [review]
LGTM

**Pull request merged**